### ISSUE TITLE
Upgrade to typeorm 0.3

### ISSUE THREAD

**Issue by annakopp** [jorgebodega/typeorm-seeding#140]

**Comment by NikPrime** [comment]
Yes, it is very important thing. I'll wait for new updates

**Comment by jorgebodega** [comment]
I have been working on this, but last weeks were really hard for me.

The problem here is that the library is relying on an internal function of TypeORM, that allow me to use existing connections instead of create/destroy them. And now that function has been deleted, Datasources does not work that way.

I'm open to any idea/pr, but the way I'll probably take is just a helper method to use on lifecycle methods to bind a Datasource.

I'm going to work on this this weekend, probably.

**Comment by a88zach** [comment]
We created a temp shim so that typeorm-seeding works with TypeORM v0.3

```
#!/usr/bin/env node

/* eslint-disable no-unused-expressions */
/* eslint-disable @typescript-eslint/no-explicit-any */
/* eslint-disable no-restricted-syntax */
import 'reflect-metadata';
import yargs from 'yargs';
import { join } from 'path';
import { importSeed, runSeeder } from 'typeorm-seeding';
import { loadFiles, importFiles } from 'typeorm-seeding/dist/utils/file.util';
import { DataSource } from 'typeorm';

class SeedCommand implements yargs.CommandModule {
  command = 'seed';

  describe = 'Runs the seeds';

  builder(args: yargs.Argv) {
    return args
      .option('dataSourcePath', {
        alias: 'd',
        describe: 'Path to your data-source file',
      })
      .option('dataSourceOptionsPath', {
        alias: 'o',
        describe: 'Path to your data-source options file',
      })
      .option('seed', {
        alias: 's',
        describe: 'Specific seed class to run.',
      })
      .demandOption(['dataSourcePath', 'dataSourceOptionsPath']);
  }

  async handler(args: yargs.Arguments) {
    const { log } = console;
    log(`🌱  TypeORM Seeding`);

    let dataSource: DataSource;
    let dataSourceOptions: { factories: string[]; seeds: string[] };
    try {
      dataSource = (
        await import(join(__dirname, '..', args.dataSourcePath as string))
      ).dataSource;
      dataSourceOptions = (
        await import(
          join(__dirname, '..', args.dataSourceOptionsPath as string)
        )
      ).dataSourceOptions;
    } catch (error) {
      console.error(error, 'Could not load the Datasource!');
      process.exit(1);
    }

    // Find all factories and seed with help of the config
    const factoryFiles = loadFiles(dataSourceOptions.factories);
    try {
      await importFiles(factoryFiles);
    } catch (error) {
      console.error(error, 'Could not import factories!');
      process.exit(1);
    }

    // Show seeds in the console
    const seedFiles = loadFiles(dataSourceOptions.seeds);
    let seedFileObjects: any[] = [];
    try {
      seedFileObjects = await Promise.all(
        seedFiles.map((seedFile) => importSeed(seedFile)),
      );
      seedFileObjects = seedFileObjects.filter(
        (seedFileObject) =>
          args.seed === undefined || args.seed === seedFileObject.name,
      );
    } catch (error) {
      console.error(error, 'Could not import seeders!');
      process.exit(1);
    }

    // Get database connection and pass it to the seeder
    try {
      await dataSource.initialize();
      (global as any).TypeORM_Seeding_Connection.connection = dataSource;
      (global as any).TypeORM_Seeding_Connection.ormConfig = dataSourceOptions;
    } catch (error) {
      console.error(
        error,
        'Database connection failed! Check your typeORM config file.',
      );
      process.exit(1);
    }

    // Run seeds
    for (const seedFileObject of seedFileObjects) {
      try {
        log(`Running seeder: ${seedFileObject.name}`);
        // eslint-disable-next-line no-await-in-loop
        await runSeeder(seedFileObject);
      } catch (error) {
        console.error(error, `Could not run the seed ${seedFileObject.name}!`);
        process.exit(1);
      }
    }

    log('👍 ', `Finished Seeding`);
    process.exit(0);
  }
}

yargs
  .usage('Usage: $0 <command> [options]')
  .command(new SeedCommand())
  .recommendCommands()
  .demandCommand(1)
  .strict()
  .help('h')
  .alias('h', 'help').argv;

```

Then in our package.json scripts
`"db:seed": "ts-node ./scripts/seed.ts seed -d $npm_package_config_typeorm_dataSourcePath -o $npm_package_config_typeorm_dataSourceOptionsPath",`

**Comment by michaelm244** [comment]
@a88zach Thanks so much for that shim! Made my setup much much easier

**Comment by adrien-may** [comment]
We did not exactly cover the same features than this lib but we implemented typeorm3 support quite recently.

@jorgebodega you can find inspiration on our repo if you need

https://www.npmjs.com/package/@adrien-may/factory
https://github.com/teamMay/factory

**Comment by mehedi-iitdu** [comment]
@jorgebodega if you need any help to upgrade to make it compatible with typeorm v3 let me know.

**Comment by jorgebodega** [comment]
I'm still working on this, please remember I can only work whenever I have free time.

Most of the logic are already migrated, but I have some troubles with seed command.

@adrien-may thanks! I'll see it.

@mehedi-iitdu I've working on `typeorm-0.3` branch, you can take a look there, and feel free to make a PR.

**Comment by jorgebodega** [comment]
Hi again!

I've been working on this today. I think most of the application is now adapted to new `Datasource` from typeorm, and probably I will release a new unstable version this weekend, so if someone want to take a look on the branch, or want to test the new version, please let me know.

**Comment by KamalAman** [comment]
@jorgebodega any luck with publishing the unstable version? I would love to be able to try it out, otherwise I could try to build from your branch on github.

**Comment by jorgebodega** [comment]
Hi! I released the version [`v5.0.0-next.1`](https://github.com/jorgebodega/typeorm-seeding/releases/tag/v5.0.0-next.1) with support for TypeORM 0.3.

I have a lot of troubles dealing with the new way of work and cli testing, so I decided to split all the factory functionality into a new package: [`typeorm-factory`](https://github.com/jorgebodega/typeorm-factory), and keep everything related with just the seeding here. That package has the same functionality, just a bit more extended. Starting on version `6` there will be no factories.

I'm going to release the actual version with deprecation notices.
